src/Microsoft.Data/PrimitiveColumn.cs

```diff
@@ -46,6 +44,36 @@ public PrimitiveColumn(string name, ReadOnlyMemory<byte> buffer, ReadOnlyMemory<
             _columnContainer = new PrimitiveColumnContainer<T>(buffer, nullBitMap, length, nullCount);
         }
 
+        /// <summary>
+        /// Returns an enumerable of mutable memory buffers representing the underlying values
+        /// </summary>
+        /// <remarks>Null values are represented as default(T)</remarks>
+        /// <returns>IEnumerable<Memory<typeparamref name="T"/>></returns>
+        public IEnumerable<Memory<T>> GetBuffers()
+        {
+            for (int i = 0; i < _columnContainer.Buffers.Count; i++)
+            {
+                DataFrameBuffer<T> buffer = DataFrameBuffer<T>.GetMutableBuffer(_columnContainer.Buffers[i]);
```

Comment:
Will people be able to check if this column has a mutable buffer before calling this method? Will they need to?

---
In reply to: [323959037](https://github.com/dotnet/corefxlab/pull/2721#discussion_r323959037) [](ancestors = 323959037)

Reply:
> Will people be able to check if this column has a mutable buffer before calling this method? Will they need to?

No APIs exist at the moment to check the mutability of the buffers. The exception being thrown here is the only indication of the mutability. I'm thinking about how to do better here. The issue is that a `PrimitiveColumn` is never immutable. If it has an immutable buffer and a setter is called, it is replaced with a mutable buffer. If I make a public API like `IsImmutable`, my concern is that it'll give the impression that a column is immutable. The more I think about it, the more I feel that users should not know about the mutability of the underlying buffers and we should do "copyOnImmutable" always:

Whenever `GetBuffers` is called, replace all immutable buffers in a column with mutable buffers and then return them. We will pay for cloning each immutable buffer, but we will avoid a potentially confusing API in return.

Reply:
I think we still have a confusing API, since someone could write:

```
var column = ...;
column.GetBuffers().First().Span[0] = newValue;
```
And after running that code, `column[0] == newValue` will sometimes be true and sometimes be false depending on how the `column` was retrieved.

What do you think about the idea to have an optional `copyOnReadOnly` option? If that parameter is `false`, throw if the buffer is readonly. If that is `true`, then do the copy. The default value would be `false`, which would throw if it was readonly.
That way the caller has to explicitly opt-in to copying.

Reply:
Couple questions:

> And after running that code, column[0] == newValue will sometimes be true and sometimes be false depending on how the column was retrieved.

Why do you say that? If we did copy on immutable always(meaning replace immutable buffers with mutable buffers in `GetBuffers()` and then return the mutable buffer), `column[0] == newValue` would always be true right? 

I'm mostly good with the `copyOnReadOnly` option. Only minor concern I have here is that the existence of the parameter implies that users would need to know if their buffers are mutable/immutable. Since the `PrimitiveColumn` is always mutable(and the buffers are `internal`, so nobody has access to them), users might wonder why they might ever 1) end up with immutable buffers(only happens during interop with Arrow) or 2) need to set copyOnReadOnly to true?

Reply:
Maybe another option is just to expose `GetReadOnlyBuffers()` for now, and wait to see if anyone needs to get mutable buffers. I think just getting the readonly buffers would cover a large set of use cases. And given the [comment below](https://github.com/dotnet/corefxlab/pull/2721#discussion_r324896974) regarding the null bitmap, maybe we should hold off on being able to mutate the buffer underneath the column until we have a solid use case for it.